<!DOCTYPE html>
<html lang="en" class="js csstransforms3d">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="generator" content="Hugo 0.74.3" />
    <meta name="description" content="ViroMatch: Viral Sequence Identification">
<meta name="author" content="T.N. Wylie">

    <link rel="icon" href="/viromatch/images/favicon.png" type="image/png">

    <title>Steps in the Pipeline :: ViroMatch</title>

    
    <link href="/viromatch/css/nucleus.css?1607305476" rel="stylesheet">
    <link href="/viromatch/css/fontawesome-all.min.css?1607305476" rel="stylesheet">
    <link href="/viromatch/css/hybrid.css?1607305476" rel="stylesheet">
    <link href="/viromatch/css/featherlight.min.css?1607305476" rel="stylesheet">
    <link href="/viromatch/css/perfect-scrollbar.min.css?1607305476" rel="stylesheet">
    <link href="/viromatch/css/auto-complete.css?1607305476" rel="stylesheet">
    <link href="/viromatch/css/atom-one-dark-reasonable.css?1607305476" rel="stylesheet">
    <link href="/viromatch/css/theme.css?1607305476" rel="stylesheet">
    <link href="/viromatch/css/hugo-theme.css?1607305476" rel="stylesheet">
    
    <link href="/viromatch/css/theme-red.css?1607305476" rel="stylesheet">
    
    

    <script src="/viromatch/js/jquery-3.3.1.min.js?1607305476"></script>

    <style>
      :root #header + #content > #left > #rlblock_left{
          display:none !important;
      }
      
        :not(pre) > code + span.copy-to-clipboard {
            display: none;
        }
      
    </style>
    
  </head>
  <body class="" data-url="/viromatch/overview/steps/">
    <nav id="sidebar" class="">



  <div id="header-wrapper">
    <div id="header">
      <a id="logo" href='/viromatch/'>




<svg version="1.1" width="200" height="50" viewBox="0 0 1100 250" xml:space="preserve" xmlns="http://www.w3.org/2000/svg">
  <g
     id="layer1">
    <g
       id="g123">
      <g
         id="g486"
         transform="matrix(8.8574711,0,0,8.242622,15.277146,-8.3433826)"
         style="fill:#ffffff">
        <g
           id="g430"
           style="fill:#ffffff">
          <path
             style="fill:#ffffff"
             d="m 31.917,17.595 c -0.911,-0.691 -2.213,-0.519 -2.91,0.396 -0.066,0.093 -0.12,0.19 -0.174,0.285 l -5.095,-0.863 c 0.048,-0.353 0.08,-0.709 0.08,-1.075 0,-1.481 -0.412,-2.863 -1.118,-4.053 l 3.248,-2.101 c 0.207,0.14 0.46,0.216 0.728,0.191 0.618,-0.057 1.072,-0.599 1.018,-1.221 -0.056,-0.618 -0.603,-1.071 -1.221,-1.016 -0.618,0.055 -1.071,0.603 -1.017,1.221 0.005,0.062 0.021,0.117 0.033,0.178 L 22.262,11.62 C 21.618,10.75 20.808,10.016 19.869,9.469 l 2.318,-3.662 c 0.247,0.04 0.509,0.003 0.742,-0.132 0.536,-0.312 0.722,-0.996 0.409,-1.532 -0.311,-0.539 -0.997,-0.725 -1.534,-0.413 -0.536,0.311 -0.721,0.997 -0.41,1.536 0.03,0.053 0.064,0.099 0.104,0.146 L 19.163,9.099 C 18.373,8.738 17.518,8.495 16.616,8.41 L 17.09,5.61 C 17.272,5.566 17.44,5.475 17.56,5.319 17.834,4.955 17.767,4.438 17.404,4.164 17.041,3.886 16.523,3.957 16.247,4.319 15.973,4.681 16.042,5.2 16.403,5.475 16.438,5.498 16.479,5.507 16.515,5.53 L 16.031,8.383 C 15.97,8.381 15.912,8.374 15.852,8.374 c -1.034,0 -2.019,0.202 -2.922,0.562 L 10.695,4.83 C 10.912,4.555 11.045,4.211 11.039,3.833 11.025,2.96 10.31,2.263 9.437,2.281 8.567,2.295 7.873,3.01 7.886,3.882 7.899,4.752 8.615,5.444 9.487,5.432 9.574,5.43 9.657,5.415 9.74,5.402 l 2.184,4.012 C 10.832,10.031 9.9,10.902 9.209,11.944 L 5.946,9.384 C 6.011,9.145 5.999,8.881 5.891,8.635 5.636,8.07 4.973,7.817 4.407,8.072 3.839,8.326 3.587,8.991 3.841,9.558 c 0.257,0.564 0.921,0.816 1.487,0.562 0.055,-0.026 0.105,-0.06 0.155,-0.092 l 3.316,2.607 c -0.49,0.937 -0.8,1.982 -0.884,3.096 L 3.02,15.722 C 2.883,15.4 2.644,15.117 2.311,14.943 1.54,14.536 0.584,14.83 0.181,15.602 c -0.405,0.771 -0.109,1.725 0.659,2.129 0.771,0.408 1.726,0.11 2.132,-0.661 0.041,-0.075 0.067,-0.153 0.095,-0.234 l 4.845,0.012 c 0.04,0.631 0.146,1.247 0.325,1.833 L 5.106,19.887 C 4.933,19.708 4.699,19.584 4.431,19.552 3.815,19.482 3.26,19.924 3.189,20.538 c -0.071,0.616 0.37,1.178 0.988,1.246 0.615,0.069 1.171,-0.374 1.241,-0.989 0.006,-0.061 0.005,-0.117 0.004,-0.182 l 3.084,-1.188 c 0.611,1.449 1.641,2.683 2.937,3.542 L 9.931,24.528 C 9.698,24.439 9.434,24.422 9.179,24.506 8.587,24.703 8.27,25.337 8.463,25.923 c 0.193,0.59 0.83,0.913 1.418,0.722 0.59,-0.198 0.908,-0.833 0.717,-1.424 -0.021,-0.058 -0.045,-0.11 -0.072,-0.167 l 1.619,-1.67 c 0.771,0.407 1.617,0.687 2.512,0.819 l -0.431,4.041 c -0.244,0.056 -0.473,0.192 -0.639,0.401 -0.38,0.491 -0.292,1.193 0.2,1.577 0.492,0.381 1.197,0.289 1.578,-0.201 0.379,-0.492 0.288,-1.195 -0.203,-1.578 -0.048,-0.033 -0.099,-0.062 -0.149,-0.093 l 0.433,-4.068 c 0.136,0.009 0.27,0.021 0.408,0.021 1.035,0 2.021,-0.203 2.93,-0.562 l 0.804,3.58 c -0.201,0.146 -0.36,0.354 -0.431,0.615 -0.16,0.6 0.193,1.215 0.794,1.376 0.599,0.157 1.216,-0.193 1.376,-0.795 0.16,-0.601 -0.196,-1.215 -0.796,-1.376 -0.059,-0.016 -0.117,-0.022 -0.178,-0.029 l -0.832,-3.706 c 0.579,-0.301 1.111,-0.671 1.597,-1.103 l 1.102,1.552 c -0.099,0.156 -0.15,0.338 -0.125,0.537 0.061,0.448 0.477,0.766 0.929,0.708 0.453,-0.063 0.768,-0.479 0.708,-0.93 -0.062,-0.453 -0.479,-0.768 -0.931,-0.707 -0.042,0.006 -0.075,0.028 -0.115,0.04 l -1.137,-1.602 c 0.838,-0.854 1.479,-1.896 1.864,-3.062 l 5.241,0.891 c 0.108,0.451 0.356,0.869 0.751,1.168 0.913,0.693 2.216,0.52 2.911,-0.395 0.692,-0.911 0.514,-2.213 -0.399,-2.908 z m -18.54,-1.626 c -0.655,0 -1.19,-0.534 -1.19,-1.194 0,-0.657 0.535,-1.188 1.19,-1.188 0.658,0 1.192,0.53 1.192,1.188 0,0.659 -0.533,1.194 -1.192,1.194 z m 3.071,2.588 c -0.482,0 -0.874,-0.392 -0.874,-0.873 0,-0.486 0.392,-0.875 0.874,-0.875 0.483,0 0.875,0.388 0.875,0.875 -0.001,0.484 -0.392,0.873 -0.875,0.873 z m 3.246,-3.703 c -0.712,0 -1.289,-0.577 -1.289,-1.287 0,-0.715 0.577,-1.293 1.289,-1.293 0.712,0 1.29,0.578 1.29,1.293 0,0.71 -0.577,1.287 -1.29,1.287 z"
             id="path428" />
        </g>
        <g
           id="g432"
           style="fill:#ffffff" />
        <g
           id="g434"
           style="fill:#ffffff" />
        <g
           id="g436"
           style="fill:#ffffff" />
        <g
           id="g438"
           style="fill:#ffffff" />
        <g
           id="g440"
           style="fill:#ffffff" />
        <g
           id="g442"
           style="fill:#ffffff" />
        <g
           id="g444"
           style="fill:#ffffff" />
        <g
           id="g446"
           style="fill:#ffffff" />
        <g
           id="g448"
           style="fill:#ffffff" />
        <g
           id="g450"
           style="fill:#ffffff" />
        <g
           id="g452"
           style="fill:#ffffff" />
        <g
           id="g454"
           style="fill:#ffffff" />
        <g
           id="g456"
           style="fill:#ffffff" />
        <g
           id="g458"
           style="fill:#ffffff" />
        <g
           id="g460"
           style="fill:#ffffff" />
      </g>
 <text x="331.8631" y="182.52051" fill="#FFFF" font-family="sans-serif" font-size="149.93px" stroke-width=".26458" style="line-height:1.25" xml:space="preserve"><tspan x="331.8631" y="182.52051" fill="#FFFFFF" font-family="'Novacento Sans Wide'" font-size="149.93px" stroke-width=".26458">ViroMatch</tspan></text>
</svg>
</a>

    </div>
    
        <div class="searchbox">
    <label for="search-by"><i class="fas fa-search"></i></label>
    <input data-search-input id="search-by" type="search" placeholder="Search...">
    <span data-search-clear=""><i class="fas fa-times"></i></span>
</div>

<script type="text/javascript" src="/viromatch/js/lunr.min.js?1607305476"></script>
<script type="text/javascript" src="/viromatch/js/auto-complete.js?1607305476"></script>
<script type="text/javascript">
    
        var baseurl = "https:\/\/twylie.github.io\/viromatch\/";
    
</script>
<script type="text/javascript" src="/viromatch/js/search.js?1607305476"></script>

    
  </div>
  
    <section id="homelinks">
      <ul>
        <li>
            <a class="padding" href='/viromatch/'><i class='fas fa-home'></i>&nbsp;Home</a>
        </li>
      </ul>
    </section>
  

    <div class="highlightable">
    <ul class="topics">

        
          
          




 
  
    
    <li data-nav-id="/viromatch/quick_start/" title="Quick Start" class="dd-item
        
        
        
        ">
      <a href="/viromatch/quick_start/">
          <b>1. </b>Quick Start
          
      </a>
      
      
        <ul>
          
          
            
          
          

        
          
            
            




 
  
    
    <li data-nav-id="/viromatch/quick_start/install_and_execution/" title="Installation and Pipeline Execution" class="dd-item
        
        
        
        ">
      <a href="/viromatch/quick_start/install_and_execution/">
          Installation and Pipeline Execution
          
      </a>
      
      
    </li>
  
 

            
          
        
        </ul>
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/viromatch/download_and_install/" title="Download and Installation" class="dd-item
        
        
        
        ">
      <a href="/viromatch/download_and_install/">
          <b>2. </b>Download and Installation
          
      </a>
      
      
        <ul>
          
          
            
          
          

        
          
            
            




 
  
    
    <li data-nav-id="/viromatch/download_and_install/prerequisites/" title="Prerequisites" class="dd-item
        
        
        
        ">
      <a href="/viromatch/download_and_install/prerequisites/">
          Prerequisites
          
      </a>
      
      
    </li>
  
 

            
          
            
            




 
  
    
    <li data-nav-id="/viromatch/download_and_install/docker/" title="Docker" class="dd-item
        
        
        
        ">
      <a href="/viromatch/download_and_install/docker/">
          Docker
          
      </a>
      
      
    </li>
  
 

            
          
            
            




 
  
    
    <li data-nav-id="/viromatch/download_and_install/download_viromatch/" title="Download ViroMatch" class="dd-item
        
        
        
        ">
      <a href="/viromatch/download_and_install/download_viromatch/">
          Download ViroMatch
          
      </a>
      
      
    </li>
  
 

            
          
            
            




 
  
    
    <li data-nav-id="/viromatch/download_and_install/databases/" title="Databases" class="dd-item
        
        
        
        ">
      <a href="/viromatch/download_and_install/databases/">
          Databases
          
      </a>
      
      
    </li>
  
 

            
          
        
        </ul>
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/viromatch/execution/" title="Running ViroMatch" class="dd-item
        
        
        
        ">
      <a href="/viromatch/execution/">
          <b>3. </b>Running ViroMatch
          
      </a>
      
      
        <ul>
          
          
            
          
          

        
          
            
            




 
  
    
    <li data-nav-id="/viromatch/execution/mapping_volumes/" title="Mapping Docker Volumes" class="dd-item
        
        
        
        ">
      <a href="/viromatch/execution/mapping_volumes/">
          Mapping Docker Volumes
          
      </a>
      
      
    </li>
  
 

            
          
            
            




 
  
    
    <li data-nav-id="/viromatch/execution/interactive_docker/" title="Interactive Docker" class="dd-item
        
        
        
        ">
      <a href="/viromatch/execution/interactive_docker/">
          Interactive Docker
          
      </a>
      
      
    </li>
  
 

            
          
            
            




 
  
    
    <li data-nav-id="/viromatch/execution/cli_docker/" title="Command Line Docker" class="dd-item
        
        
        
        ">
      <a href="/viromatch/execution/cli_docker/">
          Command Line Docker
          
      </a>
      
      
    </li>
  
 

            
          
            
            




 
  
    
    <li data-nav-id="/viromatch/execution/cli_args/" title="Command Line Arguments" class="dd-item
        
        
        
        ">
      <a href="/viromatch/execution/cli_args/">
          Command Line Arguments
          
      </a>
      
      
    </li>
  
 

            
          
        
        </ul>
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/viromatch/overview/" title="Pipeline Overview" class="dd-item
        parent
        
        
        ">
      <a href="/viromatch/overview/">
          <b>4. </b>Pipeline Overview
          
      </a>
      
      
        <ul>
          
          
            
          
          

        
          
            
            




 
  
    
    <li data-nav-id="/viromatch/overview/summary/" title="Pipeline Summary" class="dd-item
        
        
        
        ">
      <a href="/viromatch/overview/summary/">
          Pipeline Summary
          
      </a>
      
      
    </li>
  
 

            
          
            
            




 
  
    
    <li data-nav-id="/viromatch/overview/snakemake/" title="Snakemake" class="dd-item
        
        
        
        ">
      <a href="/viromatch/overview/snakemake/">
          Snakemake
          
      </a>
      
      
    </li>
  
 

            
          
            
            




 
  
    
    <li data-nav-id="/viromatch/overview/steps/" title="Steps in the Pipeline" class="dd-item
        parent
        active
        
        ">
      <a href="/viromatch/overview/steps/">
          Steps in the Pipeline
          
      </a>
      
      
    </li>
  
 

            
          
            
            




 
  
    
    <li data-nav-id="/viromatch/overview/tools/" title="Third-Party Tools" class="dd-item
        
        
        
        ">
      <a href="/viromatch/overview/tools/">
          Third-Party Tools
          
      </a>
      
      
    </li>
  
 

            
          
            
            




 
  
    
    <li data-nav-id="/viromatch/overview/file_types/" title="Input File Types" class="dd-item
        
        
        
        ">
      <a href="/viromatch/overview/file_types/">
          Input File Types
          
      </a>
      
      
    </li>
  
 

            
          
            
            




 
  
    
    <li data-nav-id="/viromatch/overview/best_hit_logic/" title="Best Hit Logic" class="dd-item
        
        
        
        ">
      <a href="/viromatch/overview/best_hit_logic/">
          Best Hit Logic
          
      </a>
      
      
    </li>
  
 

            
          
            
            




 
  
    
    <li data-nav-id="/viromatch/overview/reports/" title="Reports and Output Files" class="dd-item
        
        
        
        ">
      <a href="/viromatch/overview/reports/">
          Reports and Output Files
          
      </a>
      
      
    </li>
  
 

            
          
        
        </ul>
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/viromatch/project_info/" title="Project Info" class="dd-item
        
        
        
        ">
      <a href="/viromatch/project_info/">
          <b>5. </b>Project Info
          
      </a>
      
      
        <ul>
          
          
            
          
          

        
          
            
            




 
  
    
    <li data-nav-id="/viromatch/project_info/version/" title="Version" class="dd-item
        
        
        
        ">
      <a href="/viromatch/project_info/version/">
          Version
          
      </a>
      
      
    </li>
  
 

            
          
            
            




 
  
    
    <li data-nav-id="/viromatch/project_info/authors/" title="Authors" class="dd-item
        
        
        
        ">
      <a href="/viromatch/project_info/authors/">
          Authors
          
      </a>
      
      
    </li>
  
 

            
          
            
            




 
  
    
    <li data-nav-id="/viromatch/project_info/license/" title="License" class="dd-item
        
        
        
        ">
      <a href="/viromatch/project_info/license/">
          License
          
      </a>
      
      
    </li>
  
 

            
          
            
            




 
  
    
    <li data-nav-id="/viromatch/project_info/funding/" title="Funding" class="dd-item
        
        
        
        ">
      <a href="/viromatch/project_info/funding/">
          Funding
          
      </a>
      
      
    </li>
  
 

            
          
            
            




 
  
    
    <li data-nav-id="/viromatch/project_info/disclaimer/" title="Disclaimer" class="dd-item
        
        
        
        ">
      <a href="/viromatch/project_info/disclaimer/">
          Disclaimer
          
      </a>
      
      
    </li>
  
 

            
          
        
        </ul>
      
    </li>
  
 

          
        
    </ul>

    
    

    
    <section id="footer">
      

<p><u>Funding</u>: NIH CTSA UL1TR002345, NIH 1R01HD095986</p>

<br>

<p><u>Disclaimer</u>: This resource is intended for purely research purposes. It should not be used for emergencies or medical or professional advice.</p>

<br>

<p><a rel="license" href="http://creativecommons.org/licenses/by-nd/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nd/4.0/80x15.png" /></a><br />Site content licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nd/4.0/">Creative Commons Attribution-NoDerivatives 4.0 International License</a>. Copyright (C) 2019&mdash;2020 T.N. Wylie and K.M. Wylie.</P>

    </section>
  </div>
</nav>




        <section id="body">
        <div id="overlay"></div>
        <div class="padding highlightable">
              
              <div>
                <div id="top-bar">
                
                
                <div id="breadcrumbs" itemscope="" itemtype="http://data-vocabulary.org/Breadcrumb">
                    <span id="sidebar-toggle-span">
                        <a href="#" id="sidebar-toggle" data-sidebar-toggle="">
                          <i class="fas fa-bars"></i>
                        </a>
                    </span>
                  
                  <span id="toc-menu"><i class="fas fa-list-alt"></i></span>
                  
                  <span class="links">
                 
                 
                    
          
          
            
            
          
          
            
            
          
          
            <a href='/viromatch/'>Home</a> > <a href='/viromatch/overview/'>Pipeline Overview</a> > Steps in the Pipeline
          
        
          
        
          
        
                 
                  </span>
                </div>
                
                    <div class="progress">
    <div class="wrapper">
<nav id="TableOfContents">
  <ul>
    <li><a href="#viromatch-rulegraph-of-steps">ViroMatch Rulegraph of Steps</a></li>
    <li><a href="#steps-snakemake-rules">Steps (Snakemake Rules)</a>
      <ul>
        <li><a href="#1-prep_fastq_files">1. prep_fastq_files</a></li>
        <li><a href="#2-trim_fastq_files">2. trim_fastq_files</a></li>
        <li><a href="#3-blank_eval_filter_low_complexity">3. blank_eval_filter_low_complexity</a></li>
        <li><a href="#4-filter_low_complexity_fastq_file">4. filter_low_complexity_fastq_file</a></li>
        <li><a href="#5-host_screen_mapping">5. host_screen_mapping</a></li>
        <li><a href="#6-host_screen_write_unmapped_bam">6. host_screen_write_unmapped_bam</a></li>
        <li><a href="#7-host_screen_write_unmapped_fastq">7. host_screen_write_unmapped_fastq</a></li>
        <li><a href="#8-viral_nuc_mapping">8. viral_nuc_mapping</a></li>
        <li><a href="#9-viral_nuc_write_unmapped_bam">9. viral_nuc_write_unmapped_bam</a></li>
        <li><a href="#10-viral_nuc_write_unmapped_fastq">10. viral_nuc_write_unmapped_fastq</a></li>
        <li><a href="#11-viral_nuc_write_mapped_bam">11. viral_nuc_write_mapped_bam</a></li>
        <li><a href="#12-viral_nuc_write_mapped_fastq">12. viral_nuc_write_mapped_fastq</a></li>
        <li><a href="#13-blank_eval_viral_trans_nuc">13. blank_eval_viral_trans_nuc</a></li>
        <li><a href="#14-viral_trans_nuc_mapping">14. viral_trans_nuc_mapping</a></li>
        <li><a href="#15-viral_trans_nuc_daa_to_tsv">15. viral_trans_nuc_daa_to_tsv</a></li>
        <li><a href="#16-viral_trans_nuc_extract_mapped_ids">16. viral_trans_nuc_extract_mapped_ids</a></li>
        <li><a href="#17-viral_trans_nuc_write_mapped_fastq">17. viral_trans_nuc_write_mapped_fastq</a></li>
        <li><a href="#18-viral_mapped_fastq_merge">18. viral_mapped_fastq_merge</a></li>
        <li><a href="#19-validate_nuc_nt_mapping">19. validate_nuc_nt_mapping</a></li>
        <li><a href="#20-validate_nuc_nt_write_mapped_sam">20. validate_nuc_nt_write_mapped_sam</a></li>
        <li><a href="#21-validate_nuc_nt_write_unmapped_sam">21. validate_nuc_nt_write_unmapped_sam</a></li>
        <li><a href="#22-validate_nuc_nt_merge_r1_mapped_sam">22. validate_nuc_nt_merge_r1_mapped_sam</a></li>
        <li><a href="#23-validate_nuc_nt_merge_r2_mapped_sam">23. validate_nuc_nt_merge_r2_mapped_sam</a></li>
        <li><a href="#24-validate_nuc_nt_write_r1_unmapped_ids">24. validate_nuc_nt_write_r1_unmapped_ids</a></li>
        <li><a href="#25-validate_nuc_nt_write_r2_unmapped_ids">25. validate_nuc_nt_write_r2_unmapped_ids</a></li>
        <li><a href="#26-validate_nuc_nt_write_merged_unmapped_fastq">26. validate_nuc_nt_write_merged_unmapped_fastq</a></li>
        <li><a href="#27-blank_eval_validate_trans_nuc_nr">27. blank_eval_validate_trans_nuc_nr</a></li>
        <li><a href="#28-validate_trans_nuc_nr_mapping">28. validate_trans_nuc_nr_mapping</a></li>
        <li><a href="#29-validate_trans_nuc_nr_mapping_daa_to_tsv">29. validate_trans_nuc_nr_mapping_daa_to_tsv</a></li>
        <li><a href="#30-validate_trans_nuc_nr_merge_r1_mapped_tsv">30. validate_trans_nuc_nr_merge_r1_mapped_tsv</a></li>
        <li><a href="#31-validate_trans_nuc_nr_merge_r2_mapped_tsv">31. validate_trans_nuc_nr_merge_r2_mapped_tsv</a></li>
        <li><a href="#32-nuc_nt_best_hit_filter_sam">32. nuc_nt_best_hit_filter_sam</a></li>
        <li><a href="#33-trans_nuc_nr_best_hit_filter_tsv">33. trans_nuc_nr_best_hit_filter_tsv</a></li>
        <li><a href="#34-nuc_nt_otherseq_hit_report">34. nuc_nt_otherseq_hit_report</a></li>
        <li><a href="#35-nuc_nt_unknown_hit_report">35. nuc_nt_unknown_hit_report</a></li>
        <li><a href="#36-trans_nuc_nr_otherseq_hit_report">36. trans_nuc_nr_otherseq_hit_report</a></li>
        <li><a href="#37-trans_nuc_nr_unknown_hit_report">37. trans_nuc_nr_unknown_hit_report</a></li>
        <li><a href="#38-nuc_nt_best_hit_count_prep">38. nuc_nt_best_hit_count_prep</a></li>
        <li><a href="#39-trans_nuc_nr_best_hit_count_prep">39. trans_nuc_nr_best_hit_count_prep</a></li>
        <li><a href="#40-nuc_nt_best_hit_counts">40. nuc_nt_best_hit_counts</a></li>
        <li><a href="#41-trans_nuc_nr_best_hit_counts">41. trans_nuc_nr_best_hit_counts</a></li>
        <li><a href="#42-copy_nuc_nt_report">42. copy_nuc_nt_report</a></li>
        <li><a href="#43-copy_trans_nuc_nr_report">43. copy_trans_nuc_nr_report</a></li>
        <li><a href="#44-copy_nuc_ambiguous_report">44. copy_nuc_ambiguous_report</a></li>
        <li><a href="#45-copy_trans_nuc_ambiguous_report">45. copy_trans_nuc_ambiguous_report</a></li>
      </ul>
    </li>
  </ul>
</nav>
    </div>
</div>

                
              </div>
            </div>
            
        <div id="head-tags">
        
        </div>
        
        <div id="body-inner">
          
            <h1>
              
              Steps in the Pipeline
            </h1>
          

        



	<h2 id="viromatch-rulegraph-of-steps">ViroMatch Rulegraph of Steps</h2>
<p>The following rulegraph was produced by Snakemake and shows the general directed acyclic graph for the steps in the pipeline. Steps are described in detail below.</p>
<p><img src="rulegraph.png" alt="ViroMatch Rulegraph of Steps"></p>

<a  href="rulegraph.r47.pdf" target="_blank"  class="btn btn-default">
  
  
  
    
  
  Download PDF version of rulegraph.
  
  <i class="fas fa-download"></i>
  
</a>

<h2 id="steps-snakemake-rules">Steps (Snakemake Rules)</h2>
<p>Here is a list of all of the steps/rules in ViroMatch&rsquo;s Snakefile, ordered by occurence; each step of the pipeline and its purpose is outlined in detail.</p>
<h3 id="1-prep_fastq_files">1. prep_fastq_files</h3>
<p>The pipeline accepts a uBAM file or paired FASTQ files as <code>--input</code> from the command line. This step prepares the input file for downstream processing. If a uBAM file is supplied, it is converted to paired FASTQ files. Paired FASTQ input files are simply linked into the output directory for processing.</p>
<h3 id="2-trim_fastq_files">2. trim_fastq_files</h3>
<p>The paired FASTQ files are adaptor trimmed in this step. Reads will be trimmed for adaptor (<code>--adaptor</code>), sequences will be trimmed/revised based on end-quality (<code>--endqual</code>), reads will be filtered based on the percent of N&rsquo;s allowed post-trimming (<code>--minn</code>), and too-short reads will be removed (<code>--readlen</code>).</p>
<h3 id="3-blank_eval_filter_low_complexity">3. blank_eval_filter_low_complexity</h3>
<p>This step evaluates the resultant FASTQ files produced by the <code>trim_fastq_files</code> step to see if the FASTQ files are blank or have sequence entries present. Some of the executables in the pipeline will not accept &ldquo;blank&rdquo; FASTQ files, so this step prepares the next step in the pipeline if the files have no entries.</p>
<h3 id="4-filter_low_complexity_fastq_file">4. filter_low_complexity_fastq_file</h3>
<p>The trimmed FASTQ files are evaluated for low-complexity in this step. Low-complexity bases are &ldquo;hard masked&rdquo; as N&rsquo;s in this step. Reads will also be filtered based on the percent of Ns allowed post-trimming (<code>--minn</code>).</p>
<h3 id="5-host_screen_mapping">5. host_screen_mapping</h3>
<p>The trimmed and masked paired FASTQ sequences are mapped by nucleotide alignment to a reference host (<code>--host</code>) genome in this step.</p>
<h3 id="6-host_screen_write_unmapped_bam">6. host_screen_write_unmapped_bam</h3>
<p>For the paired FASTQ files, the unmapped reads are partitioned into their own BAM files.</p>
<h3 id="7-host_screen_write_unmapped_fastq">7. host_screen_write_unmapped_fastq</h3>
<p>The unmapped host BAM files are converted to FASTQ format.</p>
<h3 id="8-viral_nuc_mapping">8. viral_nuc_mapping</h3>
<p>The unmapped host FASTQ files (reads trimmed, masked, and did not map to host genome) are mapped by nucleotide alignment to a virus-only reference database in this step of the pipeline. Reads that map to this database have putative viral identities, but will undergo validation in later steps of the pipeline.</p>
<h3 id="9-viral_nuc_write_unmapped_bam">9. viral_nuc_write_unmapped_bam</h3>
<p>Reads that were not mapped to the viral-only database are partitioned into their own BAM files.</p>
<h3 id="10-viral_nuc_write_unmapped_fastq">10. viral_nuc_write_unmapped_fastq</h3>
<p>Reads that were not mapped to the viral-only database are written to their own FASTQ files, from the unmapped BAM files.</p>
<h3 id="11-viral_nuc_write_mapped_bam">11. viral_nuc_write_mapped_bam</h3>
<p>Reads that were mapped to the viral-only database are partitioned into their own BAM files.</p>
<h3 id="12-viral_nuc_write_mapped_fastq">12. viral_nuc_write_mapped_fastq</h3>
<p>Reads that were mapped to the viral-only database are written to their own FASTQ files, from the mapped BAM files.</p>
<h3 id="13-blank_eval_viral_trans_nuc">13. blank_eval_viral_trans_nuc</h3>
<p>This step evaluates the resultant FASTQ files produced by the viral nucleotide mapping steps to see if the FASTQ files are blank or have sequence entries present. Some of the executables in the pipeline will not accept &ldquo;blank&rdquo; FASTQ files, so this step prepares the next step in the pipeline if the files have no entries.</p>
<h3 id="14-viral_trans_nuc_mapping">14. viral_trans_nuc_mapping</h3>
<p>The reads (unmapped FASTQ files) that did not map to the viral-only databases by nucleotide alignment are aligned to a six-frame translated version of the same viral-only database by translated nucleotide alignment. Reads that did not map before may align in this manner given the greater sensitivity of the translated alignment approach.</p>
<h3 id="15-viral_trans_nuc_daa_to_tsv">15. viral_trans_nuc_daa_to_tsv</h3>
<p>We convert the translated alignment results from their binary format to a human readable tab-delimited format.</p>
<h3 id="16-viral_trans_nuc_extract_mapped_ids">16. viral_trans_nuc_extract_mapped_ids</h3>
<p>We require the read ids in list format for the reads that mapped by translated alignment. This step collects the read ids for translated alignment mapped reads.</p>
<h3 id="17-viral_trans_nuc_write_mapped_fastq">17. viral_trans_nuc_write_mapped_fastq</h3>
<p>Based on the list of read ids for translated alignment mapped reads, we can write FASTQ files for just the reads that were mapped by translated alignment.</p>
<h3 id="18-viral_mapped_fastq_merge">18. viral_mapped_fastq_merge</h3>
<p>In this step, we merge all of the reads that were mapped by (1) nucleotide alignment and (2) translated nucleotide alignment. Cumulatively, these reads are representative of all of the putative viral identity reads from the original sequences input into the pipeline; however, these candidate reads must be further validated by the pipeline.</p>
<h3 id="19-validate_nuc_nt_mapping">19. validate_nuc_nt_mapping</h3>
<p>The candidate viral reads in the merged FASTQ files are now mapped by nucleotide alignment to all of the references in the NCBI nt database. While a read may have a good hit to the viral-only reference database, it may have a better or equally good hit to a non-viral reference in NCBI nt. Aligning to NCBI nt would be prohibitively slow for all input reads, so we only do this validation step for candidate reads that aligned to the virus-only database. NCBI nt is too large to index for alignment, so we have split nt into multiple indexed reference databases for independent alignment. A candidate read will be aligned to <em>all</em> of the split nt databases and a best-hit will be determined by the pipeline downstream.</p>
<h3 id="20-validate_nuc_nt_write_mapped_sam">20. validate_nuc_nt_write_mapped_sam</h3>
<p>All of the split nt alignment BAM files are converted to SAM format for the mapped reads in this step.</p>
<h3 id="21-validate_nuc_nt_write_unmapped_sam">21. validate_nuc_nt_write_unmapped_sam</h3>
<p>All of the split nt alignment BAM files are converted to SAM format for the unmapped reads in this step.</p>
<h3 id="22-validate_nuc_nt_merge_r1_mapped_sam">22. validate_nuc_nt_merge_r1_mapped_sam</h3>
<p>For the read-1 FASTQ reads, all of the split nt SAM files for the mapped reads are merged into a single SAM file that is sorted by read id names. The SAM file needs to be sorted by read names because the pipeline&rsquo;s downstreamn best hit logic evaluates reads by <em>read blocks</em>: blocks of hits for the same read id.</p>
<h3 id="23-validate_nuc_nt_merge_r2_mapped_sam">23. validate_nuc_nt_merge_r2_mapped_sam</h3>
<p>For the read-2 FASTQ reads, all of the split nt SAM files for the mapped reads are merged into a single SAM file that is sorted by read id names. The SAM file needs to be sorted by read names because the pipeline&rsquo;s downstreamn best hit logic evaluates reads by <em>read blocks</em>: blocks of hits for the same read id.</p>
<h3 id="24-validate_nuc_nt_write_r1_unmapped_ids">24. validate_nuc_nt_write_r1_unmapped_ids</h3>
<p>For the read-1 FASTQ reads, we require the read ids in list format for the candidate viral reads that were not mapped to NCBI nt. This step collects the read ids for NCBI nt unmapped reads.</p>
<h3 id="25-validate_nuc_nt_write_r2_unmapped_ids">25. validate_nuc_nt_write_r2_unmapped_ids</h3>
<p>For the read-2 FASTQ reads, we require the read ids in list format for the candidate viral reads that were not mapped to NCBI nt. This step collects the read ids for NCBI nt unmapped reads.</p>
<h3 id="26-validate_nuc_nt_write_merged_unmapped_fastq">26. validate_nuc_nt_write_merged_unmapped_fastq</h3>
<p>In this step, based on the unmapped read ids collected in previous steps, we pull the candidate reads that did not map to NCBI nt references and merge them into FASTQ files, one for each respective original read pair.</p>
<h3 id="27-blank_eval_validate_trans_nuc_nr">27. blank_eval_validate_trans_nuc_nr</h3>
<p>This step evaluates the resultant FASTQ files produced by the NCBI nt mapping steps to see if the FASTQ files are blank or have sequence entries present. Some of the executables in the pipeline will not accept &ldquo;blank&rdquo; FASTQ files, so this step prepares the next step in the pipeline if the files have no entries.</p>
<h3 id="28-validate_trans_nuc_nr_mapping">28. validate_trans_nuc_nr_mapping</h3>
<p>The FASTQ files (from step <a href="https://twylie.github.io/viromatch/overview/steps/#26-validate_nuc_nt_write_merged_unmapped_fastq">#26</a>) contain the viral candidate reads that did not map to the NCBI nt database by nucleotide alignment. These reads will now be mapped to NCBI nr by translated nucleotide alignment. Reads that did not map before may align in this manner given the greater sensitivity of the translated alignment approach. Aligning to NCBI nr would be prohibitively slow for all input reads, so we only do this validation step for candidate reads that initially aligned to the virus-only database, but did not align to NCBI nt. NCBI nr is too large to index for alignment, so we have split nr into multiple indexed reference databases for independent alignment. A candidate read will be aligned to <em>all</em> of the split nr databases and a best-hit will be determined by the pipeline downstream.</p>
<h3 id="29-validate_trans_nuc_nr_mapping_daa_to_tsv">29. validate_trans_nuc_nr_mapping_daa_to_tsv</h3>
<p>We convert the translated alignment results from their binary format to a human readable tab-delimited format.</p>
<h3 id="30-validate_trans_nuc_nr_merge_r1_mapped_tsv">30. validate_trans_nuc_nr_merge_r1_mapped_tsv</h3>
<p>For the read-1 FASTQ reads, all of the split nr TSV files for the mapped reads are merged into a single TSV file that is sorted by read id names. The TSV file needs to be sorted by read names because the pipeline&rsquo;s downstreamn best hit logic evaluates reads by <em>read blocks</em>: blocks of hits for the same read id.</p>
<h3 id="31-validate_trans_nuc_nr_merge_r2_mapped_tsv">31. validate_trans_nuc_nr_merge_r2_mapped_tsv</h3>
<p>For the read-2 FASTQ reads, all of the split nr TSV files for the mapped reads are merged into a single TSV file that is sorted by read id names. The TSV file needs to be sorted by read names because the pipeline&rsquo;s downstreamn best hit logic evaluates reads by <em>read blocks</em>: blocks of hits for the same read id.</p>
<h3 id="32-nuc_nt_best_hit_filter_sam">32. nuc_nt_best_hit_filter_sam</h3>
<p>In this step, the candidate viral reads that mapped to NCBI nt are evaluated. For each candidate read, all of the associated hits (read block) for the read are evaluated and a best representative hit is chosen. It is during this step that the pipeline writes sanity files which describe the pass/fail status for all of the hits evaluated. A SAM file is written for the chosen viral best hits.</p>
<h3 id="33-trans_nuc_nr_best_hit_filter_tsv">33. trans_nuc_nr_best_hit_filter_tsv</h3>
<p>In this step, the candidate viral reads that mapped to NCBI nr are evaluated. For each candidate read, all of the associated hits (read block) for the read are evaluated and a best representative hit is chosen. It is during this step that the pipeline writes sanity files which describe the pass/fail status for all of the hits evaluated. A TSV file is written for the chosen viral best hits.</p>
<h3 id="34-nuc_nt_otherseq_hit_report">34. nuc_nt_otherseq_hit_report</h3>
<p>The pipeline now generates a report for hits to <em>other sequences</em> taxonomy based on the sanity reports from step <a href="https://twylie.github.io/viromatch/overview/steps/#32-nuc_nt_best_hit_filter_sam">#32</a>.</p>
<h3 id="35-nuc_nt_unknown_hit_report">35. nuc_nt_unknown_hit_report</h3>
<p>The pipeline now generates a report for hits to <em>unknown</em> taxonomy based on the sanity reports from step <a href="https://twylie.github.io/viromatch/overview/steps/#32-nuc_nt_best_hit_filter_sam">#32</a>.</p>
<h3 id="36-trans_nuc_nr_otherseq_hit_report">36. trans_nuc_nr_otherseq_hit_report</h3>
<p>The pipeline now generates a report for hits to <em>other sequences</em> taxonomy based on the sanity reports from step #33.</p>
<h3 id="37-trans_nuc_nr_unknown_hit_report">37. trans_nuc_nr_unknown_hit_report</h3>
<p>The pipeline now generates a report for hits to <em>unknown</em> taxonomy based on the sanity reports from step <a href="https://twylie.github.io/viromatch/overview/steps/#33-trans_nuc_nr_best_hit_filter_tsv">#33</a>.</p>
<h3 id="38-nuc_nt_best_hit_count_prep">38. nuc_nt_best_hit_count_prep</h3>
<p>In this step, the nucleotide alignment SAM files from step <a href="https://twylie.github.io/viromatch/overview/steps/#32-nuc_nt_best_hit_filter_sam">#32</a> that contain the best viral hits against NCBI nt are combined into a data matrix that includes associated taxonomy for reporting purposes.</p>
<h3 id="39-trans_nuc_nr_best_hit_count_prep">39. trans_nuc_nr_best_hit_count_prep</h3>
<p>In this step, the translated nucleotide alignment TSV files from step #33 that contain the best viral hits against NCBI nt are combined into a data matrix that includes associated taxonomy for reporting purposes.</p>
<h3 id="40-nuc_nt_best_hit_counts">40. nuc_nt_best_hit_counts</h3>
<p>This step generates a report file that describes viral read counts based on nucleotide alignments at genus, species, and full lineage levels.</p>
<h3 id="41-trans_nuc_nr_best_hit_counts">41. trans_nuc_nr_best_hit_counts</h3>
<p>This step generates a report file that describes viral read counts based on translated nucleotide alignments at genus, species, and full lineage levels.</p>
<h3 id="42-copy_nuc_nt_report">42. copy_nuc_nt_report</h3>
<p>The report from step <a href="https://twylie.github.io/viromatch/overview/steps/#40-nuc_nt_best_hit_counts">#40</a> is copied into the main output directory for review.</p>
<h3 id="43-copy_trans_nuc_nr_report">43. copy_trans_nuc_nr_report</h3>
<p>The report from step <a href="https://twylie.github.io/viromatch/overview/steps/#41-trans_nuc_nr_best_hit_counts">#41</a> is copied into the main output directory for review.</p>
<h3 id="44-copy_nuc_ambiguous_report">44. copy_nuc_ambiguous_report</h3>
<p>The report from steps <a href="https://twylie.github.io/viromatch/overview/steps/#34-nuc_nt_otherseq_hit_report">#34</a> and <a href="https://twylie.github.io/viromatch/overview/steps/#35-nuc_nt_unknown_hit_report">#35</a> are copied into the main output directory for review.</p>
<h3 id="45-copy_trans_nuc_ambiguous_report">45. copy_trans_nuc_ambiguous_report</h3>
<p>The report from steps <a href="https://twylie.github.io/viromatch/overview/steps/#36-trans_nuc_nr_otherseq_hit_report">#36</a> and <a href="https://twylie.github.io/viromatch/overview/steps/#37-trans_nuc_nr_unknown_hit_report">#37</a> are copied into the main output directory for review.</p>





<footer class=" footline" >
	
</footer>

        
        </div>
        

      </div>

    <div id="navigation">
        
        

        
            
            
                
                    
                    
                
                

                    
                    
                        
                    
                    

                    
                        
            
            
                
                    
                        
                        
                    
                
                

                    
                    
                        
                    
                    

                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                        
                    
                    

                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                        
                    
                    

                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        
                    
                        
            
            
                
                    
                    
                
                

                    
                    
                        
                    
                    

                    
                        
            
            
                
                    
                        
                        
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                        
                        
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                        
                    
                    

                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        
                    
            
        

        


	 
	 
		
			<a class="nav nav-prev" href="/viromatch/overview/snakemake/" title="Snakemake"> <i class="fa fa-chevron-left"></i></a>
		
		
			<a class="nav nav-next" href="/viromatch/overview/tools/" title="Third-Party Tools" style="margin-right: 0px;"><i class="fa fa-chevron-right"></i></a>
		
	
    </div>

    </section>

    <div style="left: -1000px; overflow: scroll; position: absolute; top: -1000px; border: none; box-sizing: content-box; height: 200px; margin: 0px; padding: 0px; width: 200px;">
      <div style="border: none; box-sizing: content-box; height: 200px; margin: 0px; padding: 0px; width: 200px;"></div>
    </div>
    <script src="/viromatch/js/clipboard.min.js?1607305476"></script>
    <script src="/viromatch/js/perfect-scrollbar.min.js?1607305476"></script>
    <script src="/viromatch/js/perfect-scrollbar.jquery.min.js?1607305476"></script>
    <script src="/viromatch/js/jquery.sticky.js?1607305476"></script>
    <script src="/viromatch/js/featherlight.min.js?1607305476"></script>
    <script src="/viromatch/js/highlight.pack.js?1607305476"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <script src="/viromatch/js/modernizr.custom-3.6.0.js?1607305476"></script>
    <script src="/viromatch/js/learn.js?1607305476"></script>
    <script src="/viromatch/js/hugo-learn.js?1607305476"></script>
    
    

  </body>
</html>
